when building a layered docker image, ignore it if tar encounters cha…

purefn opened the pull request:
###### Motivation for this change
Allows building layered docker images in parallel with other derivations. Takes the second approach mentioned in #75888, telling tar to ignore the changes to `/nix/store`. `tar` exit codes are defined as

```
       0      Successful termination.

       1      Some files differ.  If tar was invoked with the --compare (--diff, -d) command line option, this means that some files in the  archive  differ
              from  their  disk counterparts.  If tar was given one of the --create, --append or --update options, this exit code means that some files were
              changed while being archived and so the resulting archive does not contain the exact copy of the file set.

       2      Fatal error.  This means that some fatal, unrecoverable error occurred.
```
So as long as `tar` exits with a `0` or `1` we should be fine. The rationale for that is that any derivations that building the layered docker image depends on would have already been built and in the nix store by the time we get to this point. Therefore, any changes to `/nix/store` don't pertain to the docker image under construction.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @nlewo

jonringer commented:
cc @grahamc

purefn commented:
Hmm I'm still seeing an intermittent error 

```
docker-image-learning-server-nix.tar.gz> Cooking the image...
docker-image-learning-server-nix.tar.gz>  tar: ./6773e5a7a063b5386300f1675b617904d21ee4b8c7f9d75b275c67a62cf18e0a/layer.tar: file changed as we read it
```

That seems to be coming from this bit

```
echo "Cooking the image..."                                                                                                                        
tar -C image --dereference --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0  --mode=a-w --xform s:'^./':: -c . |       pigz -nT > $out       
```

I'm a bit confused. I don't see how that file could be modified at this point and I'm hesitant to give it the same treatment without better understanding.

nlewo commented:
I'm not in favor of this approach. I would prefer to explore the approach described in https://github.com/NixOS/nixpkgs/issues/75888#issuecomment-567833985.

purefn commented:
@nlewo - I've updated it with a different approach which should work better. Let me know what you think!

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,29 @@ set -eu
     --mtime="@$SOURCE_DATE_EPOCH" \
-    --owner=0 --group=0 "$@"
+    --owner=0 --group=0 \
+    --transform="s,$n,/nix/store/$n," \
```
I don't understand why this change would be required. If I remember well, tar doesn't create entries for the directories `/nix` and `/nix/store` in the tar archive. Since they are not explicitly created by tar, it doesn't read them and I don't see any issues here regarding these entries.

purefn replied:
You are correct that `tar` does not create entries for `/nix` and `/nix/store`. But even though it doesn't create those entries, it does still check if they are modified while it's running. I was able to force `tar` to fail in this way by running a `nix build` of a bunch of packages and at the same time running 

```
tar -cf test.tar /nix/store/vs7zapqwngylvpydpv5dg3gk1va8dj8w-bash-interactive-4.4-p23
```

repeatedly.

nlewo replied:
Ok. Nice stress test and thx for your explanation.

nlewo reviewed (changes requested):
Otherwise, your first change looks fine to me ;)

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,29 @@ set -eu
 mkdir -p "$layerPath"
-tar --no-recursion -rf "$layerPath/layer.tar" \
+
+# make sure /nix and /nix/store appear first in the archive
```
Could you coment we are doing this to avoid file changes during the tar process.

purefn replied:
Done.

nlewo reviewed (changes requested):
Could you reformat your commit message?
Otherwise, the code looks good to me ;)

purefn commented:
> Could you reformat your commit message?

Reformat it how? It's formatted according to the CONTRIBUTING.md unless there is a detail I'm missing.

jonringer commented:
>Reformat it how? It's formatted according to the CONTRIBUTING.md unless there is a detail I'm missing.

Since dockerTools refers to an attr set, something like:
```
dockerTools.buildLayeredImage: fix building images in parallel
```

Commit:
dockerTools.buildLayeredImage: fix building layered images in parallel

when tar'ing store paths into layered archives when building layered
images, don't use the absolute nix store path so that tar won't complain
if something new is added to the nix store

when building the final docker image, ignore any file changes tar
detects in the layers. they are all immutable and the only thing that
might change is the number of hard links due to store optimization

purefn commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,29 @@ set -eu
     --mtime="@$SOURCE_DATE_EPOCH" \
-    --owner=0 --group=0 "$@"
+    --owner=0 --group=0 \
+    --transform="s,$n,/nix/store/$n," \
```
Done.

purefn commented:
@nlewo Changed the final image tar to ignore the file changes it detected, added comments, and reformatted git commit as per @jonringer's suggestion. Should be good to merge now. Let me know if there is anything else.

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -625,7 +624,22 @@ rec {
         echo "Cooking the image..."
-        tar -C image --dereference --hard-dereference --sort=name --mtime="@$SOURCE_DATE_EPOCH" --owner=0 --group=0  --mode=a-w --xform s:'^./':: -c . | pigz -nT > $out
+        # tar exits with an exit code of 1 if files changed while it was
+        # reading them. it considers a change in the number of hard links
```
```suggestion
        # reading them. It considers a change in the number of hard links
```

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -625,7 +624,22 @@ rec {
+        # tar exits with an exit code of 1 if files changed while it was
+        # reading them. it considers a change in the number of hard links
+        # to be a "change", which can cause this to fail if images are being
+        # built concurrently and auto-optimise-store is turned on. since
```
```suggestion
        # built concurrently and the auto-optimise-store nix option is turned on. Since
```

nlewo commented on pkgs/build-support/docker/default.nix:
```diff
@@ -625,7 +624,22 @@ rec {
+        # reading them. it considers a change in the number of hard links
+        # to be a "change", which can cause this to fail if images are being
+        # built concurrently and auto-optimise-store is turned on. since
+        # know the contents of these files will not change, we can reasonably
```
```suggestion
        # the contents of these files will not change, we can reasonably
```

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,43 @@ set -eu
 mkdir -p "$layerPath"
-tar --no-recursion -rf "$layerPath/layer.tar" \
+
+# make sure /nix and /nix/store appear first in the archive.
```
```suggestion
# Make sure /nix and /nix/store appear first in the archive.
```

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,43 @@ set -eu
-tar --no-recursion -rf "$layerPath/layer.tar" \
+
+# make sure /nix and /nix/store appear first in the archive.
+# we create the directories here and use them because
```
```suggestion
# We create the directories here and use them because
```

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,43 @@ set -eu
+    --transform='s,nix,/nix,' \
+    nix
+
+# we change into the /nix/store in order to avoid a similar
```
```suggestion
# We change into the /nix/store in order to avoid a similar
```

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,43 @@ set -eu
+# we change into the /nix/store in order to avoid a similar
+# "file changed as we read it" error as above. Namely,
+# if we use the absolute path of /nix/store/123-pkg
+# and something new it added to the nix store while tar
```
```suggestion
# and something new is added to the nix store while tar
```

nlewo commented on pkgs/build-support/docker/store-path-to-layer.sh:
```diff
@@ -5,16 +5,43 @@ set -eu
+# the relative nix store path, tar will ignore changes
+# to /nix/store. In order to create the correct structure
+# in the tar file, we transform the relative nix store
+# path to the absolute store path
```
```suggestion
# path to the absolute store path.
```

nlewo reviewed (changes requested):
Otherwise, LGTM! And thanks for all of these comments:)

nlewo commented:
@GrahamcOfBorg test docker-tools

nlewo commented:
Note I applied comment fixes in da261e36316b5b32009e8aa936e32df4b91e11bd.

Thanks!

rimmington commented:
@nlewo Would it be possible to backport this fix to 19.09?

nlewo commented:
@rimmington I'm sorry, I don't have time to backport it. Note also this is actually not a trivial backport: we need to backport another patch (because this one is buggy) and there is a merge conflict.

rimmington commented:
@nlewo Ah, I didn't realise the backport is non-trivial; I may just wait for 20.03 then. Thanks for checking!
